Feat: support enums as data source

Abdulmajeed-Jamaan opened the pull request:
Introduces the DataSourceEnumContract interface for enum-based data sources. Updates Collectors and FieldsContract to support enums implementing this contract, allowing for more flexible data source definitions and improved collection logic.

- [x] update the docs with usage example

Abdulmajeed-Jamaan committed:
Add DataSourceEnumContract and enhance data source handling

Introduces the DataSourceEnumContract interface for enum-based data sources. Updates Collectors and FieldsContract to support enums implementing this contract, allowing for more flexible data source definitions and improved collection logic.

Abdulmajeed-Jamaan committed:
Fix styling

Abdulmajeed-Jamaan committed:
Improve data source handling in FieldsContract

Added support for enum-based data sources in FieldsContract, enabling mapping and joining of enum cases for field options. Also, cache is now cleared in local environments to facilitate development. Minor formatting fix in Collectors.php.

Abdulmajeed-Jamaan committed:
Merge branch 'feat/support-enums-as-datasource' of https://github.com/lara-zeus/bolt into feat/support-enums-as-datasource

Abdulmajeed-Jamaan commented:
Sample of the an enum:

```php
<?php

namespace App\Zeus\DataSources;

use Filament\Support\Contracts\HasLabel;
use Illuminate\Contracts\Support\Htmlable;
use LaraZeus\Bolt\DataSources\DataSourceEnumContract;

enum Status: string implements HasLabel, DataSourceEnumContract
{
    case Approve = 'approve';
    case Pending = 'pending';

    public function getLabel(): string|Htmlable|null
    {
        return match ($this) {
            self::Approve => 'approved',
            self::Pending => 'waiting'
        };
    }

    public function getDataSourceLabel(): string|Htmlable|null
    {
        return $this->getLabel();
    }

    public static function toDataSourceArray(): array
    {
        return [
            'title' => 'Status',
            'disabled' => false,
            'sort' => 1,
            'class' => '\\' . self::class
        ];
    }
}
```

The structure of the array returned from `toDataSourceArray` is mandatory for this feature to work, it make sure the enums blend well the the bolt internal system.

Question is: should we rely on the developer to read docs to write the required data or use DTOs to enforce them from the php level ?

Abdulmajeed-Jamaan committed:
Remove redundant null coalescing in join result

Eliminated the unnecessary null coalescing operator after the join method call, as join() always returns a string. This simplifies the code and avoids unreachable fallback logic.

Abdulmajeed-Jamaan committed:
Merge branch '4.x' into feat/support-enums-as-datasource

Abdulmajeed-Jamaan committed:
Add DataSourceData class and update enum contract

Introduces the DataSourceData class to encapsulate data source properties. Updates DataSourceEnumContract to return DataSourceData instead of an array, and refactors Collectors facade to use the new DataSourceData structure for improved type safety and clarity.

Abdulmajeed-Jamaan commented:
Updated to use Data Object as validator for required data and their types.

```php
    public static function toDataSourceData(): DataSourceData
    {
        return DataSourceData::make('Status', static::class)
            ->disabled(false) // optional
            ->sort(1); // optional
    }
```

atmonshi commented:
to be consist with the current structure, I would prefer if all contracts are in the namespace: `LaraZeus\Bolt\Contracts`
and rename `DataSourceEnumContract.php` to `DataSourceEnum.php`

did some testing, all prefect.
thank you.

Abdulmajeed-Jamaan committed:
Move DataSourceEnumContract to Contracts and rename

Renamed DataSourceEnumContract to DataSourceEnum and moved it from DataSources to Contracts namespace. Updated all references to use the new interface and namespace, improving code organization and consistency.

Abdulmajeed-Jamaan committed:
Fix styling

Abdulmajeed-Jamaan committed:
Merge branch '4.x' into feat/support-enums-as-datasource

Abdulmajeed-Jamaan committed:
Update datasource docs with enum support and caching info

Added documentation for creating data sources from enums by implementing the DataSourceEnum interface. Moved and expanded caching information, clarified customization options, and improved section organization for better readability.

atmonshi committed:
Update add-datasource.md

Abdulmajeed-Jamaan committed:
Update make() to require title and class parameters

The make() method in DataSourceData now requires explicit 'title' and 'class' arguments, improving clarity and type safety when creating new instances.
